control-base/src/main/java/fi/nls/oskari/control/feature/InsertFeatureHandler.java

```diff
@@ -35,58 +39,62 @@
+                }
 
-        final WFSLayerConfiguration wfsMetadata = getWFSConfiguration(layer.getId());
-        final String wfstMessage = createWFSTMessage(jsonPayload, layer, wfsMetadata);
-        LOG.debug("Inserting feature to service at", layer.getUrl(), "with payload", wfstMessage);
-        final String responseString = postPayload(layer, wfstMessage);
-        final String updatedFeatureId = parseFeatureIdFromResponse(responseString);
+                final String wfstMessage = createWFSTMessage(featureJSON);
+                LOG.debug("Inserting feature to service at", layer.getUrl(), "with payload", wfstMessage);
+                final String responseString = postPayload(layer, wfstMessage);
+                updatedFeatureIds.put(parseFeatureIdFromResponse(responseString));
 
-        flushLayerTilesCache(layer.getId());
-        ResponseHelper.writeResponse(params, JSONHelper.createJSONObject("fid", updatedFeatureId));
+                flushLayerTilesCache(layer.getId());
+            }
+            ResponseHelper.writeResponse(params, JSONHelper.createJSONObject("fids", updatedFeatureIds));
+        } catch (JSONException e) {
+            LOG.error(e, "JSON processing error");
+            throw new ActionException("JSON processing error", e);
+        }
     }
 
-    private String createWFSTMessage(JSONObject jsonPayload, OskariLayer layer, WFSLayerConfiguration wfsMetadata)
+    private String createWFSTMessage(JSONObject jsonPayload)
             throws ActionException {
-        final String srsName = JSONHelper.getStringFromJSON(jsonPayload, "srsName", "http://www.opengis.net/gml/srs/epsg.xml#3067");
-        // TODO: rewrite to use wfs-t related code under myplaces OR atleast using an xml lib
-        final StringBuilder requestData = new StringBuilder("<wfs:Transaction service='WFS' version='1.1.0' xmlns:ogc='http://www.opengis.net/ogc' xmlns:wfs='http://www.opengis.net/wfs'><wfs:Insert><" + layer.getName() + " xmlns:" + wfsMetadata.getFeatureNamespace() + "='" + wfsMetadata.getFeatureNamespaceURI() + "'>");
-        try {
-            final JSONArray jsonArray = jsonPayload.getJSONArray("featureFields");
-            for (int i = 0; i < jsonArray.length(); i++) {
-                String key = jsonArray.getJSONObject(i).getString("key");
-                String value = jsonArray.getJSONObject(i).getString("value");
-                if (value.isEmpty() == false) {
-                    requestData.append("<" + key + ">" + value + "</" + key + ">");
-                }
-            }
 
-            if (jsonPayload.has("geometries")) {
-                insertGeometries(wfsMetadata.getGMLGeometryProperty(), requestData, jsonPayload.getJSONObject("geometries"), srsName);
-            }
-        } catch (JSONException ex) {
-            throw new ActionParamsException("Couldn't create WFS-T message from params", jsonPayload.toString(), ex);
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
```

Comment:
Could move the ByteArrayOutputStream and the return clause to inside the try-catch block.